### PR TITLE
Add Dataset wrapper class

### PR DESCRIPTION
**Pull request by NielsRogge**
This PR aims to simplify the design of Fondant by relying on a single `Manifest` class which replaces the existing `ExpressDataset` and `ExpressDatasetDraft` classes. The manifest consists of a single index, one or more data sources (which are references to datasets stored locally/remotely) as well as metadata (which includes things like the `run_id`).

The first component (which is always a loading component) in a KubeFlow pipeline creates the `Manifest`, and each subsequent component (which is always a transform component) takes a `Manifest` as input and returns a `Manifest`.  Moreover, Fondant takes care of updating the metadata for transform components, meaning this is not an argument of the `transform` method anymore. There are 2 classes which components can inherit from; `ExpressLoaderComponent` and `ExpressTransformComponent`.

# TLDR

A loading component creates the initial manifest:
```
from express.common import ExpressLoaderComponent, Manifest
from datasets import load_dataset

class MyLoadingComponent(ExpressLoaderComponent):
    @classmethod
    def load(
        cls, args: Optional[Dict[str, Union[str, int, float, bool]]] = None, metadata=None) -> Manifest:   
            # loading a dataset from the hub for instance
            dataset = load_dataset("mnist")

            # creates the initial manifest
            manifest = Manifest(index=..., data_sources=..., metadata=metadata)

            return manifest
``` 

A transform component takes in a manifest and returns a new one:
```
from express.common import ExpressTransformComponent, Manifest

class MyTransformComponent(ExpressTransformComponent):
    @classmethod
    def transform(cls, manifest, args) -> Manifest:

         manifest.add_sources(...)
         manifest.update_index(...)

         return manifest
```

In a follow-up PR, the `Manifest` class can be abstracted away entirely for the user. This could for instance entail that we provide a single `dataframe` to the user, which joins the requested data sources from the Manifest, checks whether indices are shared, etc.

### CODE REVIEW

**Comment by GeorgesLorre** [comment]
Open question: could we start with an empty manifest (or abstract the initial manifest creation) so we can avoid having 2 types of components (ones that create manifests and ones that update it) ?

**Review by PhilippeMoussalli — COMMENTED** [review]
Thanks Niels! I left a few comments. 

Could you write down or sketch out how you propose to create the framework specific implementation of Express? It's still not clear to me how this will be with the current approach and it might make reviewing the PR a bit easier.

File: examples/pipelines/finetune_stable_diffusion/components/embedding_component/requirements.txt [hunk 1]
```diff
@@ -1,4 +1,5 @@
 torch==2.0.0
 transformers==4.27.3
-git+https://github.com/ml6team/express.git@0f43bbd88826ffd1b620100275700f9cbe7c0b6e
-Pillow==9.4.0
\ No newline at end of file
+git+https://github.com/ml6team/express.git@4f863bdf12eb29764aad1fe1195c342e54da0074
+Pillow==9.4.0
+datasets==2.10.1 # TODO  remove this, currently common.py uses datasets only
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply examples/pipelines/finetune_stable_diffusion/components/embedding_component/requirements.txt:5]: Yes also saw that, have removed the comment and included `datasets` as a dependency

[the remark this answers is not in the document]
Comment (NielsRogge) [reply examples/pipelines/finetune_stable_diffusion/components/embedding_component/requirements.txt:5]: Will use the extra package feature instead!

File: express/components/common.py
[the remark this answers is not in the document]
Comment (PhilippeMoussalli) [reply express/components/common.py:1]: and in which class/method will those components be classes have to be instantiated? Do we also still want to have a `HFDatasetTransform/HFDatasetLoad` and instantiate them there?

[the remark this answers is not in the document]
Comment (NielsRogge) [reply express/components/common.py:1]: We can include the desired framework in the component specification, and then instantiate the required framework specific class inside the `Manifest` class. There's no need for separate `HFDatasetTransform`/`HFDatasetLoad` classes, only a single `HFDatasetsHandler` class which defines the logic to load Parquet files into HuggingFace Dataset objects.

File: examples/pipelines/finetune_stable_diffusion/components/embedding_component/src/main.py [hunk 2]
```diff
@@ -47,57 +41,52 @@ def embed(examples, processor, model):
...
     Component that embeds the images using a CLIP model from Hugging Face.
     """
 
     @classmethod
     def transform(
         cls,
-        data: HFDatasetsDataset,
-        extra_args: Optional[Dict[str, Union[str, int, float, bool]]] = None,
-    ) -> HFDatasetsDatasetDraft:
+        manifest,
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply examples/pipelines/finetune_stable_diffusion/components/embedding_component/src/main.py:52]: Could you clarify this comment? Not really following what you mean here. The transform method takes 2 arguments: `args` and `manifest`.

**Comment by NielsRogge** [comment]
@RobbeSneyders agree that keeping the Manifest as a standalone dataclass makes sense. I've updated the PR to use it. The class introduced in this PR is a wrapper around it. As can be seen, the wrapper uses the dataclass when reading from and writing to JSON. Let me know what you think.

Currently the dataclass is called DataManifest and the wrapper is called Manifest, which can be confusing. Maybe let's settle for `Manifest` for the dataclass and `FondantManifest` for the wrapper class? Ultimately both classes should be hidden from the user so they are only for internal use.

Regarding

> Also, can we settle on the subsets name instead of data_sources?

Actually I'm fine with `data_sources`, since they just represent references to a remote dataset. Wondering what @PhilippeMoussalli @GeorgesLorre's opinion is

File: examples/pipelines/finetune_stable_diffusion/components/embedding_component/src/main.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (NielsRogge) [reply examples/pipelines/finetune_stable_diffusion/components/embedding_component/src/main.py]: Thanks, updated them!

**Comment by RobbeSneyders** [comment]
> @RobbeSneyders agree that keeping the Manifest as a standalone dataclass makes sense. I've updated the PR to use it. The class introduced in this PR is a wrapper around it. As can be seen, the wrapper uses the dataclass when reading from and writing to JSON. Let me know what you think.

I still think we're combining too much and would prefer to translate our discussions on the dataset standard to code step by step, starting with a class representing the manifest. We also still need to implement the column specification etc. there.

> Currently the dataclass is called DataManifest and the wrapper is called Manifest, which can be confusing. Maybe let's settle for Manifest for the dataclass and FondantManifest for the wrapper class? Ultimately both classes should be hidden from the user so they are only for internal use.

The manifest represents a dataset, so I would only use the `Manifest` name there. Once it's loaded, it becomes a dataset and the naming should probably reflect it.

> Actually I'm fine with data_sources, since they just represent references to a remote dataset. Wondering what @PhilippeMoussalli @GeorgesLorre's opinion is

Based on the discussion in the doc, `subsets` was the only one without real objection. I think `data_sources` is ambiguous with the `source` field from the index.

File: express/components/common.py
[the remark this answers is not in the document]
Comment (NielsRogge) [reply express/components/common.py:1]: So to make it more concrete:

```
class FondantManifest:
        def __init__(self, handler="pandas", index, data_sources, metadata):
                self.handler = self._create_handler(handler)
                self.index = index

        def _upload_parquet():
               self.handler.upload_parquet(...)
```

[the remark this answers is not in the document]
Comment (GeorgesLorre) [reply express/components/common.py:1]: IMO

I'm not convinced that we need different handlers (especially for a first version) we need to decide on one framework that we can leverage to achieve what we want. Fondant will (lazy-)load and combine all needed subsets in the component and expose a dataframe that is specific for the framework we choose (pandas, dask, polars,...). The goal is to offer and abstract deep functionality and optimisation and not widespread support.

We will need components that can read different kinds of data and convert and write it to parquet to be used downstream.

Happy to hear all thoughts on this.

[the remark this answers is not in the document]
Comment (RobbeSneyders) [reply express/components/common.py:1]: Agree for the first version indeed.

**Comment by GeorgesLorre** [comment]
> Based on the discussion in the doc, `subsets` was the only one without real objection. I think `data_sources` is ambiguous with the `source` field from the index.

+1 for `subsets`

File: express/components/common.py
[the remark this answers is not in the document]
Comment (PhilippeMoussalli) [reply express/components/common.py:1]: So then do we stick to handling the loading/uploading with one framework hardcoded into the `common.py`? I think it's something that will have to tackle eventually so maybe it's worthwhile implementing the logic behind it (even if we stick to one component for the moment).

[the remark this answers is not in the document]
Comment (NielsRogge) [reply express/components/common.py:1]: Ok, I'm currently leveraging HF Datasets (the `datasets` package is imported in common.py in this PR). But if we'd want to support several frameworks, then we would require additional functionality like a handler

[the remark this answers is not in the document]
Comment (GeorgesLorre) [reply express/components/common.py:1]: But it should be limited no ? Only a method for downloading and another for uploading. 

Let's say we use Polars:

1. LoaderComponent: uses HF datasets to fetch data -> returns polars dataframe
(behind the curtains, a manifest is created and the data is stored on storage)

2. FilterComponent: Supplies Polars dataframe -> filter magic -> return Polars dataframe (filtered index)
(behind the curtains, manifest is updated and the data is stored on storage)

3. EmbedComponemt: Supplies Polars dataframe -> embed magic -> return Polars dataframe (extra columns)
(behind the curtains, manifest is updated and the data is stored on storage)

[the remark this answers is not in the document]
Comment (NielsRogge) [reply express/components/common.py:1]: Yes it should be quite limited, just methods for up- and downloading

[the remark this answers is not in the document]
Comment (PhilippeMoussalli) [reply express/components/common.py:1]: **uses HF datasets to fetch data -> returns polars dataframe** I think still each component should handle/implement one framework. For this one, it could be loading the HF dataframe and writing it to a parquet. The next component can then load it in any framework. 
 
 Framework specific functionalities should only include streaming reads/writes but we still need to define a common class that handles them.
